File: DSA/string hashing/string_hashing.py
def compute_hash(s, p=31, m=10**9 + 9):
    n = len(s)
    hash_vals = [0] * (n + 1)
    p_powers = [1] * (n + 1)

    for i in range(1, n + 1):
        p_powers[i] = (p_powers[i - 1] * p) % m

    for i in range(n):
        val = ord(s[i]) - ord('a') + 1
        hash_vals[i + 1] = (hash_vals[i] + val * p_powers[i]) % m

    return hash_vals, p_powers

# Modular inverse using Fermat's Little Theorem
def mod_inverse(x, m):
    return pow(x, m - 2, m)

# Get hash of s[l:r] (0-based, inclusive)
def substring_hash(l, r, hash_vals, p_powers, m=10**9 + 9):
    raw_hash = (hash_vals[r + 1] - hash_vals[l] + m) % m
    norm = mod_inverse(p_powers[l], m)
    return (raw_hash * norm) % m

File: DSA/string hashing/string_hashing.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

typedef long long ll;
const ll P = 31;             // Prime base
const ll M = 1e9 + 9;        // Modulus for hashing

vector<ll> hash_vals, p_powers;

// Function to compute prefix hashes and powers of P
void compute_hash(const string &s, ll p = P, ll m = M) {
    int n = s.size();
    hash_vals.resize(n + 1, 0);
    p_powers.resize(n + 1, 1);

    for (int i = 1; i <= n; ++i) {
        p_powers[i] = (p_powers[i - 1] * p) % m;
    }

    for (int i = 0; i < n; ++i) {
        ll val = s[i] - 'a' + 1;  // Map 'a' -> 1, 'b' -> 2, etc.
        hash_vals[i + 1] = (hash_vals[i] + val * p_powers[i]) % m;
    }
}

// Function to compute hash of substring s[l...r] (0-based)
ll substring_hash(int l, int r, ll m = M) {
    ll raw_hash = (hash_vals[r + 1] - hash_vals[l] + m) % m;
    // Use modular inverse of p_powers[l] for correct normalization
    return (raw_hash * p_powers[p_powers.size() - 2 - l]) % m;
}

int main() {
    string s = "abcabc";
    compute_hash(s);

    // Example: hash of substring s[0...2] = "abc"
    int l = 0, r = 2;
    cout << "Hash of substring \"" << s.substr(l, r - l + 1) << "\" = " << substring_hash(l, r) << endl;

    return 0;
}
